### ISSUE TITLE
✨ Support for .gitignore

### ISSUE THREAD

**Issue by benjaminstrasser** [dagger/dagger#3736]
### What are you trying to do?

Provide an easy way to clear a directory of any files ignored by git. The formal definition can be found [here ](https://git-scm.com/docs/gitignore).

My suggestion would be the following:
`func (r *Directory) WithoutGitignore(path string) *Directory`
`This directory with all files removed that would be ignored by git`

Side Note:
I don't currently need this but `WithoutUntracked` and other exclusions based on git might be useful as well.

### Why is this important to you?

All CI platforms I know (Gitlab CI, GitHub Actions, etc..) don't have access to files ignored by git.
When running dagger locally ignored files might be present due to the local development process.
Those files can interfere and create different pipeline results.
It would be very convenient to be able to reproduce CI platform behavior with little effort.

### How are you currently working around this?

```
client.Host().Workdir().
    WithoutDirectory("node_modules").
    WithoutDirectory("apps/frontend/node_modules").
    WithoutDirectory("apps/backend/node_modules")
```

**Comment by marcosnils** [comment]
until it's decided if this will is a good candidate for the core API, you can currently do this by using the recently added `include` and `exclude` arguments in 0.4 https://github.com/dagger/dagger/pull/3583. In case you're wondering which is the pattern matching logic behind that `include` and `exclude` arguments, you can check it from here: https://github.com/moby/patternmatcher.  It basically supports `exact`, `prefix`, `suffix` and `regexp` matching. 

. The `.gitignore` native feature will probably take some time as AFAIK there's no good native `.gitignore` parsers in Go that we could probably use outside the box.

**Comment by benjaminstrasser** [comment]
I am not sure if simply parsing .gitignore files is sufficient as gitignored files are defined as followed:
> A gitignore file specifies intentionally untracked files that Git should ignore. Files already tracked by Git are not affected

Using include and exclude might also cause problems, as tracked files, which match a .gitignore pattern will be removed. This file would have been present inside any CI framwork but now iw ignored.

To fully replicate this behavior one needs:

- all .gitignore files and their content
- the .git folder and its content.

Some extra, but probably unnecessary files for now
- $GIT_DIR/info/exclude
- Patterns read from the file specified by the configuration variable core.excludesFile

I hastily threw some code together using the go SDK, which should currently achieve this although I am not sure how to proceed from here.

```
func WithoutIgnored(client *dagger.Client, ctx context.Context, dir *dagger.Directory) (_ *dagger.Directory, _ error) {
	gitcontainer := client.Container().
		From("alpine/git").
		WithMountedDirectory(".", dir).
		Exec(dagger.ContainerExecOpts{
			Args: []string{"ls-files", "--ignored", "--exclude-standard", "--others", "--directory", "--no-empty-directory"},
		})

	exitcode, _ := gitcontainer.ExitCode(ctx)
	if exitcode != 0 {
		return nil, errors.New("git exited with non-zero exit code")
	}

	contents, err := gitcontainer.Stdout().Contents(ctx)
	if err != nil {
		return nil, err
	}

	lines := strings.Split(contents, "\n")


	for _, line := range lines[:len(lines)-1] {
		if strings.HasSuffix(line, "/") {
			dir = dir.WithoutDirectory(line)
		} else {
			dir = dir.WithoutFile(line)
		}
	}

	return dir, nil
}
```

**Comment by marcosnils** [comment]
> I am not sure if simply parsing .gitignore files is sufficient as gitignored files are defined as followed:
> 
> >

I agree, that's why I didn't suggest parsing the `.gitignore` files since that wasn't going to be enough. One thing you can do though is to get the output of `git status --ignored` or `git check-ignore  **/*` and use that as an input for the `exclude` argument.

**Comment by benjaminstrasser** [comment]
As @vito said in [discord](https://discordapp.com/channels/707636530424053791/1037509195404169276) it looks like it is best to solve this issue by
> running a git command to figure out the ignored files and then exclude:ing them

I think this leads us to two follow-up questions.
- Does dagger want to provide such a feature or is it something every developer can copy/paste into their own util (Similar to the GoBuild utility used in [hello-monorepo](https://github.com/kpenfound/hello-monorepo/blob/main/daggerutils/main.go))?
- If dagger wants to provide it; Where would it be implemented(Presumably the SDK or a separate module/library)?

I would love some input and discussion on this.

**Comment by dolanor** [comment]
That's a good question.
I think it would be a great candidate for an extension (that are still not there, yet).
Otherwise, it would make us develop it in duplicate for each SDK language.

For now, until we have extension, I would suggest to create your own library of utility like that you would share in your SDK language.
Once the extensions are available, other SDKs would be able to use it as well.
